### ISSUE TITLE
JR requires mime type and parser to work

### ISSUE THREAD

**Issue by weskinner** [JSONAPI-Resources/jsonapi-resources#767]
I couldn't get JR working out of the box on rails 4.1 without the following in an initializer

```
Mime::Type.register "application/vnd.api+json", :json

ActionDispatch::ParamsParser::DEFAULT_PARSERS[Mime::Type.lookup('application/vnd.api+json')]=lambda do |body|
  JSON.parse(body)
end
```

Without the `Mime::Type.register` I get `415 Unsupported Media` Type for PATCHes.

Without the parser setup I get `400 Missing Parameter` saying it's missing the data param.

**Comment by dfjosh** [comment]
I get an error with or without `Mime::Type.register "application/vnd.api+json", :json`:

```
{"errors":[{"title":"Not acceptable","detail":"All requests must use the 'application/vnd.api+json' Accept without media type parameters. This request specified 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'.","code":"406","status":"406"}]}
```

I'm on Rails 5 and using `gem 'jsonapi-resources', github: 'cerebris/jsonapi-resources', branch: 'master'`

**Comment by lgebhardt** [comment]
@dfjosh You seem to have the accept header set to `text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'."`, which does not contain `application/vnd.api+json` or `*.*` without a media type parameter. http://jsonapi.org/format/#content-negotiation

**Comment by dfjosh** [comment]
@lgebhardt I'm requesting from a browser which I've just learned won't sent the correct header? I'm guessing this won't be an issue when I hook it up to emberjs. I'm just used to testing requests in the browser.

**Comment by lgebhardt** [comment]
It should be fine with emberjs. I use curl for simple tests:

``` bash
curl -i -H "Accept: application/vnd.api+json" "http://localhost:3000/phone-numbers
```

Chrome also has the Requestly extension https://github.com/requestly/chrome-extension if you want to make requests through the browser. Like any browser extension you do need to trust it.

**Comment by NuckChorris** [comment]
Highly recommend Paw if you're on OSX or Postman elsewhere too.  There's a lot of great options for testing APIs that don't involve a browser

**Comment by sebyx07** [comment]
create a base controller for json-api and add this. then you can inherit from it for other controllers

``` ruby
skip_before_action :ensure_correct_media_type
skip_before_action :ensure_valid_accept_media_type
```

**Comment by weskinner** [comment]
What I've ended up with is a monkey patch of ActsAsResourceController but I think this is the most correct solution as it doesn't turn my other "application/json" requests into "application/vnd.api+json".  It's all in an initializer mime_types.rb:

```
api_mime_types = %W(
  application/vnd.api+json
  text/x-json
  application/json
)

Mime::Type.unregister :json
Mime::Type.register 'application/json', :json, api_mime_types

ActionDispatch::ParamsParser::DEFAULT_PARSERS[Mime::Type.lookup('application/vnd.api+json')]=lambda do |body|
  JSON.parse(body)
end

module JSONAPI
  module ActsAsResourceController
    def ensure_correct_media_type
      unless request.format.json?
        fail JSONAPI::Exceptions::UnsupportedMediaTypeError.new(request.content_type)
      end
    rescue => e
      handle_exceptions(e)
    end
  end
end
```

**Comment by allanberry** [comment]
@weskinner I tried your mime_types.rb solution, and I got an error: 

`/Users/aljabear/Projects/visualist/config/initializers/mime_types.rb:15:in '<top (required)>': uninitialized constant ActionDispatch::ParamsParser (NameError)`

Rails 5.1, JR 0.9.0

Did this ever get otherwise solved?

**Comment by bf4** [comment]
@allanberry you're on Rails 5.1 which is different and well supported, while OP is on 4.1. 

Perhaps related to replacement of mime-types gem with mini-mime?

**Comment by allanberry** [comment]
@bf4 not sure, could be.  I since migrated to [FastJSONAPI](https://github.com/Netflix/fast_jsonapi); so unavailable to test anymore.  :/   Thanks for your help regardless

**Comment by ghost** [comment]
> What I've ended up with is a monkey patch of ActsAsResourceController but I think this is the most correct solution as it doesn't turn my other "application/json" requests into "application/vnd.api+json". It's all in an initializer mime_types.rb:
> 
> ```
> api_mime_types = %W(
>   application/vnd.api+json
>   text/x-json
>   application/json
> )
> 
> Mime::Type.unregister :json
> Mime::Type.register 'application/json', :json, api_mime_types
> 
> ActionDispatch::ParamsParser::DEFAULT_PARSERS[Mime::Type.lookup('application/vnd.api+json')]=lambda do |body|
>   JSON.parse(body)
> end
> 
> module JSONAPI
>   module ActsAsResourceController
>     def ensure_correct_media_type
>       unless request.format.json?
>         fail JSONAPI::Exceptions::UnsupportedMediaTypeError.new(request.content_type)
>       end
>     rescue => e
>       handle_exceptions(e)
>     end
>   end
> end
> ```

@weskinner Do you know why this solution don't work in POST/PUT requests? Here's an example:
```
curl -i -H "Accept: application/vnd.api+json" -H "Content-Type:application/vnd.api+json" -X POST -d '{"data": {"type": "users","attributes": {"name": "Foo"}}}' http://localhost:3000/api/v1/users

{
  "errors": [{
    "title":"Unsupported media type",
    "detail":"All requests that create or update must use the 'application/vnd.api+json' Content-Type. This request specified 'application/json'.",
    "code":"415",
    "status":"415"
  }]
}
```
For GET requests the @weskinner's monkey patch works well. Anyone could help?
Thanks!

**Comment by bf4** [comment]
@lgebhardt This issue should be closed. It's been resolved in JR for some time and is attracting off-topic and incorrect solutions.

tl;dr if you're not using the "application/vnd.api+json" mime type, you're doing it wrong. *don't* register "application/vnd.api+json" as 'json'. 

Rails lets you specify the mime type (aka the format) in the URL. So, in a browser, you can browse to `v1/users.jsonapi` and it'll treat the request as jsonapi.  Some older versions of JR might require the mime type be called api_json. As a developer, it should be pretty easy to figure out which version of JR you're on and the name it uses to register the "application/vnd.api+json" mime type

**Comment by lgebhardt** [comment]
@bf4 good point
